Shader compilation errors on Metal (macos 10.13.6)

I'm trying to run [Veloren](https://veloren.net/) on Mac, and I'm getting the following naga errors:

```
❯ RUST_BACKTRACE=full ./airshipper
[WARN] Naga generated shader:
// language: metal2.0
#include <metal_stdlib>
#include <simd/simd.h>

struct Globals {
    metal::float4x4 transform;
};
struct VertexInput {
    metal::float2 v_pos;
    metal::float2 pos;
    metal::float2 scale;
    metal::float2 atlas_pos;
    metal::float2 atlas_scale;
    int layer;
};
struct VertexOutput {
    metal::float4 position;
    metal::float2 uv;
    float layer;
};

struct vs_mainInput {
    metal::float2 v_pos [[attribute(0)]];
    metal::float2 pos [[attribute(1)]];
    metal::float2 scale [[attribute(2)]];
    metal::float2 atlas_pos [[attribute(3)]];
    metal::float2 atlas_scale [[attribute(4)]];
    int layer [[attribute(5)]];
};
struct vs_mainOutput {
    metal::float4 position [[position]];
    metal::float2 uv [[user(loc0), center_perspective]];
    float layer [[user(loc1), center_perspective]];
};
vertex vs_mainOutput vs_main(
  vs_mainInput varyings [[stage_in]]
, constant Globals& globals [[buffer(0)]]
) {
    const VertexInput input = { varyings.v_pos, varyings.pos, varyings.scale, varyings.atlas_pos, varyings.atlas_scale, varyings.layer };
    VertexOutput out;
    metal::float4x4 transform;
    out.uv = static_cast<float2>((input.v_pos * input.atlas_scale) + input.atlas_pos);
    out.layer = static_cast<float>(input.layer);
    transform = metal::float4x4(metal::float4(input.scale.x, 0.0, 0.0, 0.0), metal::float4(0.0, input.scale.y, 0.0, 0.0), metal::float4(0.0, 0.0, 1.0, 0.0), metal::float4(input.pos, 0.0, 1.0));
    metal::float4x4 _e40 = globals.transform;
    metal::float4x4 _e41 = transform;
    out.position = (_e40 * _e41) * metal::float4(input.v_pos, 0.0, 1.0);
    VertexOutput _e48 = out;
    const auto _tmp = _e48;
    return vs_mainOutput { _tmp.position, _tmp.uv, _tmp.layer };
}


struct fs_mainInput {
    metal::float2 uv [[user(loc0), center_perspective]];
    float layer [[user(loc1), center_perspective]];
};
struct fs_mainOutput {
    metal::float4 member1 [[color(0)]];
};
fragment fs_mainOutput fs_main(
  fs_mainInput varyings1 [[stage_in]]
, metal::float4 position [[position]]
, metal::sampler u_sampler [[sampler(0)]]
, metal::texture2d_array<float, metal::access::sample> u_texture [[texture(0)]]
) {
    const VertexOutput input1 = { position, varyings1.uv, varyings1.layer };
    metal::float4 _e7 = u_texture.sample(u_sampler, input1.uv, static_cast<int>(input1.layer));
    return fs_mainOutput { _e7 };
}

[WARN] Shader translation error for stage VERTEX: Metal: Compilation failed:

program_source:42:26: error: unknown type name 'float2'; did you mean 'float'?
    out.uv = static_cast<float2>((input.v_pos * input.atlas_scale) + input.atlas_pos);
                         ^~~~~~
                         float
program_source:42:14: error: static_cast from 'metal::float2' (aka 'vector_float2') to 'float' is not allowed
    out.uv = static_cast<float2>((input.v_pos * input.atlas_scale) + input.atlas_pos);
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

[WARN] Please report it to https://github.com/gfx-rs/naga
[ERROR] wgpu error: Validation Error

Caused by:
    In Device::create_render_pipeline
      note: label = `iced_wgpu::image pipeline`
    Internal error in VERTEX shader: Metal: Compilation failed:

program_source:42:26: error: unknown type name 'float2'; did you mean 'float'?
    out.uv = static_cast<float2>((input.v_pos * input.atlas_scale) + input.atlas_pos);
                         ^~~~~~
                         float
program_source:42:14: error: static_cast from 'metal::float2' (aka 'vector_float2') to 'float' is not allowed
    out.uv = static_cast<float2>((input.v_pos * input.atlas_scale) + input.atlas_pos);
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~



[ERROR] Airshipper panicked:

Handling wgpu errors as fatal by default: panicked at 'Handling wgpu errors as fatal by default', /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.10.1/src/backend/direct.rs:2159:5
thread 'main' panicked at 'Handling wgpu errors as fatal by default', /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.10.1/src/backend/direct.rs:2159:5
stack backtrace:
   0:        0x10681b7f2 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h8e4ba4b55e218172
   1:        0x1064ed63c - core::fmt::write::h678ac65252cdb936
   2:        0x106819b0a - std::io::Write::write_fmt::h6e4183435968e8c5
   3:        0x10681aa57 - std::panicking::default_hook::{{closure}}::h2d4327b072954705
   4:        0x106819ea9 - std::panicking::default_hook::hf452eeb666b7d536
   5:        0x106428759 - airshipper::error::panic_hook::{{closure}}::h4a2dd3a01392fef2
   6:        0x1068199fe - std::panicking::rust_panic_with_hook::hde5e2d5c940e2374
   7:        0x1068a626d - std::panicking::begin_panic::{{closure}}::hc5aa8631e6df413a
   8:        0x1068a6238 - std::sys_common::backtrace::__rust_end_short_backtrace::hb08e1bb139842c60
   9:        0x1069ac3bd - std::panicking::begin_panic::hc1881e98836881de
  10:        0x1068b0bc4 - wgpu::backend::direct::default_error_handler::h329ce89e2898eebe
  11:        0x1068b0a10 - core::ops::function::Fn::call::h019d29cd931cd9d4
  12:        0x1068c8374 - wgpu::Device::create_render_pipeline::h2c1b2a6aa17fdc59
  13:        0x106596728 - iced_wgpu::backend::Backend::new::hbebfe5b87e416756
  14:        0x10645714a - airshipper::gui::run::h18e0829aef36590d
  15:        0x10641bb46 - airshipper::cli::process::hb418343c5d05897f
  16:        0x1063b0c36 - airshipper::main::h5ef89ad357ef1e7c
  17:        0x1063ee92a - std::sys_common::backtrace::__rust_begin_short_backtrace::hfa8db7b546750383
  18:        0x1063b15c4 - _main
zsh: abort      RUST_BACKTRACE=full ./airshipper
```

--- a/src/back/msl/writer.rs
+++ b/src/back/msl/writer.rs
@@ -1194,11 +1194,9 @@ impl<W: Write> Writer<W> {
                 convert,
             } => {
                 let scalar = scalar_kind_string(kind);
-                let (size, width) = match *context.resolve_type(expr) {
-                    crate::TypeInner::Scalar { width, .. } => ("", width),
-                    crate::TypeInner::Vector { size, width, .. } => {
-                        (back::vector_size_str(size), width)
-                    }
+                let width = match *context.resolve_type(expr) {
+                    crate::TypeInner::Scalar { width, .. }
+                    | crate::TypeInner::Vector { width, .. } => width,
                     _ => return Err(Error::Validation),
                 };
                 let op = match convert {
@@ -1209,7 +1207,22 @@ impl<W: Write> Writer<W> {
                     Some(_) => return Err(Error::Validation),
                     None => "as_type",
                 };
-                write!(self.out, "{}<{}{}>(", op, scalar, size)?;
+                write!(self.out, "{}<", op)?;
+                match *context.resolve_type(expr) {
+                    crate::TypeInner::Vector { size, .. } => {
+                        write!(
+                            self.out,
+                            "{}::{}{}",
+                            NAMESPACE,
+                            scalar,
+                            back::vector_size_str(size)
+                        )?;
+                    }
+                    _ => {
+                        write!(self.out, "{}", scalar)?;
+                    }
+                }
+                write!(self.out, ">(")?;
                 self.put_expression(expr, context, true)?;
                 write!(self.out, ")")?;
             }

--- a/tests/in/operators.wgsl
+++ b/tests/in/operators.wgsl
@@ -16,8 +16,10 @@ fn builtins() -> vec4<f32> {
     // bitcast()
     let b1 = bitcast<f32>(v_i32_one.x);
     let b2 = bitcast<vec4<f32>>(v_i32_one);
+    // convert
+    let v_i32_zero = vec4<i32>(v_f32_zero);
     // done
-    return vec4<f32>(vec4<i32>(s1)) + s2 + m1 + m2 + b1 + b2;
+    return vec4<f32>(vec4<i32>(s1) + v_i32_zero) + s2 + m1 + m2 + b1 + b2;
 }
 
 fn splat() -> vec4<f32> {

--- a/tests/out/glsl/operators.main.Compute.glsl
+++ b/tests/out/glsl/operators.main.Compute.glsl
@@ -19,7 +19,8 @@ vec4 builtins() {
     vec4 m2_ = mix(vec4(0.0, 0.0, 0.0, 0.0), vec4(1.0, 1.0, 1.0, 1.0), 0.10000000149011612);
     float b1_ = intBitsToFloat(ivec4(1, 1, 1, 1).x);
     vec4 b2_ = intBitsToFloat(ivec4(1, 1, 1, 1));
-    return (((((vec4(ivec4(s1_)) + s2_) + m1_) + m2_) + vec4(b1_)) + b2_);
+    ivec4 v_i32_zero = ivec4(vec4(0.0, 0.0, 0.0, 0.0));
+    return (((((vec4((ivec4(s1_) + v_i32_zero)) + s2_) + m1_) + m2_) + vec4(b1_)) + b2_);
 }
 
 vec4 splat() {

--- a/tests/out/hlsl/operators.hlsl
+++ b/tests/out/hlsl/operators.hlsl
@@ -17,7 +17,8 @@ float4 builtins()
     float4 m2_ = lerp(float4(0.0, 0.0, 0.0, 0.0), float4(1.0, 1.0, 1.0, 1.0), 0.10000000149011612);
     float b1_ = float(int4(1, 1, 1, 1).x);
     float4 b2_ = float4(int4(1, 1, 1, 1));
-    return (((((float4(int4(s1_.xxxx)) + s2_) + m1_) + m2_) + float4(b1_.xxxx)) + b2_);
+    int4 v_i32_zero = int4(float4(0.0, 0.0, 0.0, 0.0));
+    return (((((float4((int4(s1_.xxxx) + v_i32_zero)) + s2_) + m1_) + m2_) + float4(b1_.xxxx)) + b2_);
 }
 
 float4 splat()

--- a/tests/out/msl/access.msl
+++ b/tests/out/msl/access.msl
@@ -46,7 +46,7 @@ vertex fooOutput foo(
     for(int _i=0; _i<5; ++_i) c.inner[_i] = type9 {a, static_cast<int>(b), 3, 4, 5}.inner[_i];
     c.inner[vi + 1u] = 42;
     int value = c.inner[vi];
-    return fooOutput { matrix * static_cast<float4>(metal::int4(value)) };
+    return fooOutput { matrix * static_cast<metal::float4>(metal::int4(value)) };
 }
 
 

--- a/tests/out/msl/image.msl
+++ b/tests/out/msl/image.msl
@@ -16,7 +16,7 @@ kernel void main1(
 , metal::texture1d<uint, metal::access::write> image_dst [[user(fake0)]]
 ) {
     metal::int2 dim = int2(image_storage_src.get_width(), image_storage_src.get_height());
-    metal::int2 itc = (dim * static_cast<int2>(local_id.xy)) % metal::int2(10, 20);
+    metal::int2 itc = (dim * static_cast<metal::int2>(local_id.xy)) % metal::int2(10, 20);
     metal::uint4 value1_ = image_mipmapped_src.read(metal::uint2(itc), static_cast<int>(local_id.z));
     metal::uint4 value2_ = image_multisampled_src.read(metal::uint2(itc), static_cast<int>(local_id.z));
     float value3_ = image_depth_multisampled_src.read(metal::uint2(itc), static_cast<int>(local_id.z));

--- a/tests/out/msl/operators.msl
+++ b/tests/out/msl/operators.msl
@@ -19,15 +19,16 @@ metal::float4 builtins(
     metal::float4 m1_ = metal::mix(v_f32_zero, v_f32_one, v_f32_half);
     metal::float4 m2_ = metal::mix(v_f32_zero, v_f32_one, 0.10000000149011612);
     float b1_ = as_type<float>(v_i32_one.x);
-    metal::float4 b2_ = as_type<float4>(v_i32_one);
-    return ((((static_cast<float4>(metal::int4(s1_)) + s2_) + m1_) + m2_) + metal::float4(b1_)) + b2_;
+    metal::float4 b2_ = as_type<metal::float4>(v_i32_one);
+    metal::int4 v_i32_zero = static_cast<metal::int4>(v_f32_zero);
+    return ((((static_cast<metal::float4>(metal::int4(s1_) + v_i32_zero) + s2_) + m1_) + m2_) + metal::float4(b1_)) + b2_;
 }
 
 metal::float4 splat(
 ) {
     metal::float2 a = ((metal::float2(1.0) + metal::float2(2.0)) - metal::float2(3.0)) / metal::float2(4.0);
     metal::int4 b = metal::int4(5) % metal::int4(2);
-    return a.xyxy + static_cast<float4>(b);
+    return a.xyxy + static_cast<metal::float4>(b);
 }
 
 int unary(

--- a/tests/out/spv/operators.spvasm
+++ b/tests/out/spv/operators.spvasm
@@ -1,12 +1,12 @@
 ; SPIR-V
 ; Version: 1.0
 ; Generator: rspirv
-; Bound: 99
+; Bound: 101
 OpCapability Shader
 %1 = OpExtInstImport "GLSL.std.450"
 OpMemoryModel Logical GLSL450
-OpEntryPoint GLCompute %92 "main"
-OpExecutionMode %92 LocalSize 1 1 1
+OpEntryPoint GLCompute %94 "main"
+OpExecutionMode %94 LocalSize 1 1 1
 OpMemberDecorate %22 0 Offset 0
 OpMemberDecorate %22 1 Offset 16
 %2 = OpTypeVoid
@@ -35,15 +35,15 @@ OpMemberDecorate %22 1 Offset 16
 %25 = OpConstantComposite  %19  %6 %6 %6 %6
 %26 = OpConstantComposite  %20  %7 %7 %7 %7
 %29 = OpTypeFunction %19
-%53 = OpTypeVector %4 2
-%69 = OpTypeFunction %8
-%76 = OpConstantNull  %8
-%78 = OpTypePointer Function %22
-%81 = OpTypeFunction %4
-%85 = OpTypePointer Function %19
-%87 = OpTypeInt 32 0
-%86 = OpConstant  %87  0
-%93 = OpTypeFunction %2
+%55 = OpTypeVector %4 2
+%71 = OpTypeFunction %8
+%78 = OpConstantNull  %8
+%80 = OpTypePointer Function %22
+%83 = OpTypeFunction %4
+%87 = OpTypePointer Function %19
+%89 = OpTypeInt 32 0
+%88 = OpConstant  %89  0
+%95 = OpTypeFunction %2
 %28 = OpFunction  %19  None %29
 %27 = OpLabel
 OpBranch %30
@@ -59,70 +59,72 @@ OpBranch %30
 %39 = OpCompositeExtract  %8  %26 0
 %40 = OpBitcast  %4  %39
 %41 = OpBitcast  %19  %26
-%42 = OpCompositeConstruct  %20  %31 %31 %31 %31
-%43 = OpConvertSToF  %19  %42
-%44 = OpFAdd  %19  %43 %32
-%45 = OpFAdd  %19  %44 %36
-%46 = OpFAdd  %19  %45 %37
-%47 = OpCompositeConstruct  %19  %40 %40 %40 %40
-%48 = OpFAdd  %19  %46 %47
-%49 = OpFAdd  %19  %48 %41
-OpReturnValue %49
+%42 = OpConvertFToS  %20  %24
+%43 = OpCompositeConstruct  %20  %31 %31 %31 %31
+%44 = OpIAdd  %20  %43 %42
+%45 = OpConvertSToF  %19  %44
+%46 = OpFAdd  %19  %45 %32
+%47 = OpFAdd  %19  %46 %36
+%48 = OpFAdd  %19  %47 %37
+%49 = OpCompositeConstruct  %19  %40 %40 %40 %40
+%50 = OpFAdd  %19  %48 %49
+%51 = OpFAdd  %19  %50 %41
+OpReturnValue %51
 OpFunctionEnd
-%51 = OpFunction  %19  None %29
-%50 = OpLabel
-OpBranch %52
+%53 = OpFunction  %19  None %29
 %52 = OpLabel
-%54 = OpCompositeConstruct  %53  %14 %14
-%55 = OpCompositeConstruct  %53  %3 %3
-%56 = OpFAdd  %53  %55 %54
-%57 = OpCompositeConstruct  %53  %15 %15
-%58 = OpFSub  %53  %56 %57
-%59 = OpCompositeConstruct  %53  %16 %16
-%60 = OpFDiv  %53  %58 %59
-%61 = OpCompositeConstruct  %20  %17 %17 %17 %17
-%62 = OpCompositeConstruct  %20  %18 %18 %18 %18
-%63 = OpSMod  %20  %61 %62
-%64 = OpVectorShuffle  %19  %60 %60 0 1 0 1
-%65 = OpConvertSToF  %19  %63
-%66 = OpFAdd  %19  %64 %65
-OpReturnValue %66
+OpBranch %54
+%54 = OpLabel
+%56 = OpCompositeConstruct  %55  %14 %14
+%57 = OpCompositeConstruct  %55  %3 %3
+%58 = OpFAdd  %55  %57 %56
+%59 = OpCompositeConstruct  %55  %15 %15
+%60 = OpFSub  %55  %58 %59
+%61 = OpCompositeConstruct  %55  %16 %16
+%62 = OpFDiv  %55  %60 %61
+%63 = OpCompositeConstruct  %20  %17 %17 %17 %17
+%64 = OpCompositeConstruct  %20  %18 %18 %18 %18
+%65 = OpSMod  %20  %63 %64
+%66 = OpVectorShuffle  %19  %62 %62 0 1 0 1
+%67 = OpConvertSToF  %19  %65
+%68 = OpFAdd  %19  %66 %67
+OpReturnValue %68
 OpFunctionEnd
-%68 = OpFunction  %8  None %69
-%67 = OpLabel
-OpBranch %70
-%70 = OpLabel
-%71 = OpLogicalNot  %10  %9
-OpSelectionMerge %72 None
-OpBranchConditional %71 %73 %74
-%73 = OpLabel
-OpReturnValue %7
-%74 = OpLabel
-%75 = OpNot  %8  %7
-OpReturnValue %75
+%70 = OpFunction  %8  None %71
+%69 = OpLabel
+OpBranch %72
 %72 = OpLabel
-OpReturnValue %76
+%73 = OpLogicalNot  %10  %9
+OpSelectionMerge %74 None
+OpBranchConditional %73 %75 %76
+%75 = OpLabel
+OpReturnValue %7
+%76 = OpLabel
+%77 = OpNot  %8  %7
+OpReturnValue %77
+%74 = OpLabel
+OpReturnValue %78
 OpFunctionEnd
-%80 = OpFunction  %4  None %81
-%79 = OpLabel
-%77 = OpVariable  %78  Function
-OpBranch %82
-%82 = OpLabel
-%83 = OpCompositeConstruct  %19  %3 %3 %3 %3
-%84 = OpCompositeConstruct  %22  %83 %7
-OpStore %77 %84
-%88 = OpAccessChain  %85  %77 %86
-%89 = OpLoad  %19  %88
-%90 = OpCompositeExtract  %4  %89 0
-OpReturnValue %90
+%82 = OpFunction  %4  None %83
+%81 = OpLabel
+%79 = OpVariable  %80  Function
+OpBranch %84
+%84 = OpLabel
+%85 = OpCompositeConstruct  %19  %3 %3 %3 %3
+%86 = OpCompositeConstruct  %22  %85 %7
+OpStore %79 %86
+%90 = OpAccessChain  %87  %79 %88
+%91 = OpLoad  %19  %90
+%92 = OpCompositeExtract  %4  %91 0
+OpReturnValue %92
 OpFunctionEnd
-%92 = OpFunction  %2  None %93
-%91 = OpLabel
-OpBranch %94
-%94 = OpLabel
-%95 = OpFunctionCall  %19  %28
-%96 = OpFunctionCall  %19  %51
-%97 = OpFunctionCall  %8  %68
-%98 = OpFunctionCall  %4  %80
+%94 = OpFunction  %2  None %95
+%93 = OpLabel
+OpBranch %96
+%96 = OpLabel
+%97 = OpFunctionCall  %19  %28
+%98 = OpFunctionCall  %19  %53
+%99 = OpFunctionCall  %8  %70
+%100 = OpFunctionCall  %4  %82
 OpReturn
 OpFunctionEnd

--- a/tests/out/wgsl/operators.wgsl
+++ b/tests/out/wgsl/operators.wgsl
@@ -15,7 +15,8 @@ fn builtins() -> vec4<f32> {
     let m2_: vec4<f32> = mix(vec4<f32>(0.0, 0.0, 0.0, 0.0), vec4<f32>(1.0, 1.0, 1.0, 1.0), 0.10000000149011612);
     let b1_: f32 = bitcast<f32>(vec4<i32>(1, 1, 1, 1).x);
     let b2_: vec4<f32> = vec4<f32>(vec4<i32>(1, 1, 1, 1));
-    return (((((vec4<f32>(vec4<i32>(s1_)) + s2_) + m1_) + m2_) + vec4<f32>(b1_)) + b2_);
+    let v_i32_zero: vec4<i32> = vec4<i32>(vec4<f32>(0.0, 0.0, 0.0, 0.0));
+    return (((((vec4<f32>((vec4<i32>(s1_) + v_i32_zero)) + s2_) + m1_) + m2_) + vec4<f32>(b1_)) + b2_);
 }
 
 fn splat() -> vec4<f32> {

